### PR TITLE
node: Consolidate borsch library usage

### PR DESCRIPTION
**Pull request by djb15**
I spent some time looking through the existing tests and adding a couple of new ones to check whether we're hitting all the edge cases we care about. Not a place we want to be making a mistake, so I'm happy to abandon these changes if the risks outweigh the upside

### CODE REVIEW

File: node/pkg/watchers/solana/client_test.go [hunk 1]
```diff
@@ -267,7 +264,29 @@ func TestParseMessagePublicationAccount(t *testing.T) {
 				Sequence:         3458072,
 				EmitterChain:     1,
 				EmitterAddress:   emitterAddrUnreliable,
-				Payload:          nil, // borsh deserialization results in this being nil rather than an empty slice
+				Payload:          nil, // Borsh deserialization results in this being nil rather than an empty slice
+			},
+			errStr: "",
+		},
+		{
+			name: "success -- trailing bytes are ignored",
```
Comment (johnsaigle) [inline node/pkg/watchers/solana/client_test.go]: This is a change in behaviour, right? Previously I believe this returned an error, and that's probably safer

Comment (johnsaigle) [reply node/pkg/watchers/solana/client_test.go]: If gagliardetto doesn't do this then I think that's arguably a bug in that library

Comment (johnsaigle) [reply node/pkg/watchers/solana/client_test.go]: Got this from an LLM FWIW:

---

_Borsh rejects trailing bytes. The reference implementation (borsh-rs) treats any unconsumed bytes after deserialization as an error.
The high-level entry points enforce this:
- from_slice / try_from_slice — after T::deserialize, check if !v_mut.is_empty() and return Err(InvalidData, "Not all bytes read") (borsh-rs/borsh/src/de/mod.rs).
- from_reader / try_from_reader — after deserializing, attempt to read 1 more byte; success (non-EOF) triggers the same "Not all bytes read" error.
The low-level deserialize / deserialize_reader methods themselves do not check — they only consume what the type needs. So the "no trailing bytes" rule is enforced at the API boundary, not per-field. The written spec (the README pseudocode) doesn't call it out explicitly, but the reference implementation's behavior is the de facto standard, and the error constant ERROR_NOT_ALL_BYTES_READ makes the stance unambiguous: the entire input must be consumed._

Comment (djb15) [reply node/pkg/watchers/solana/client_test.go]: Yes definitely a change of behaviour, although I couldn't think of a way to abuse this, and I also couldn't think of a neat way to keep the behaviour without just creating a wrapper function with the extra trailing bytes validation step. Any strong opinions (including abandoning this PR)?

Comment (johnsaigle) [reply node/pkg/watchers/solana/client_test.go]: I think the PR is good and I do think it would be better to make just such a wrapper function! I'd prefer that the Guardian follows the expected behaviour of Borsh even if it's not exploitable at the moment. It'll keep things easier to reason about IMO, plus set us up to be more compatible with other libraries.

It's also possible that the solana-go behaviour has been patched in a newer version, so if we upgrade that library in the future, that might address it also.

I think it'd be best if there was a code comment on top of it where we link the Borsh standard. 

This would be easy to do in a follow-up PR instead, but it would mean that we'd have to revert these changes to the test behaviour later.

**Review by bemic — COMMENTED** [review]
I haven't really looked into "shim" code. Therefore, consider this comment more as an datapoint that you can take into account when making the decision. 

I diff tested `near/borsh-go v0.3.0` vs `gagliardetto/binary v0.7.7` across all four affected struct layouts (`PostMessageData`, `MessagePublicationAccount`, `ShimPostMessageData`, `ShimMessageEventData`) using  more than 2.5mil canonical, edge case and malformed inputs. No difference in accepted inputs or successfully decoded values, and no panic.

Only difference spot by gpt5.6:

> The meaningful differences are that gagliardetto/binary decodes directly into the destination, so failed decoding may leave partial data and reused structs may retain an old slice when a zero-length payload is decoded. It also returns []byte slices that alias the input buffer, whereas near/borsh-go copied them; error messages and wrapping differ as well.
> 
> None of these differences currently breaks the reviewed flows because they use fresh destinations, discard values on error, and do not mutate decoded payloads.

**Comment by johnsaigle** [comment]
@djb15 needs rebase

File: node/pkg/watchers/solana/client_test.go (hunk 1 above)
Comment (djb15) [reply node/pkg/watchers/solana/client_test.go]: I updated this to enforce that there are never trailing bytes when borsch decoding now in the latest commit

File: node/pkg/watchers/solana/shim.go [hunk 2]
```diff
@@ -116,7 +115,7 @@ func shimParsePostMessage(shimPostMessageDiscriminator []byte, buf []byte) (*Shi
 	}
 
 	data := new(ShimPostMessageData)
-	if err := borsh.Deserialize(data, buf[len(shimPostMessageDiscriminator):]); err != nil {
+	if err := decodeBorsh(data, buf[len(shimPostMessageDiscriminator):]); err != nil {
```
Comment (johnsaigle) [inline node/pkg/watchers/solana/shim.go]: Hey I followed up on the Borsh trailing bytes thing. It looks like Rust actually does it in two different ways, allowing trailing bytes in some cases and giving an error in other cases.

Also, the [Shim post_message function on-chain](https://github.com/wormhole-foundation/wormhole/blob/e155608982af3a6e0813dafc54ecba3933559e3c/svm/wormhole-core-shims/crates/shim/src/post_message.rs#L157-L181) explicitly says that trailing bytes are ignored.

So my understanding right now is that the strict checks that you added based on my earlier review in `decodeBorsh` are generally the right way to go, especially for decoding single values. However, for decoding shim post message data _specifically_, we should allow trailing bytes so that we don't accidentally give an error here. 

I believe the change now would just be to replace this line with a call to `Decode` with no `HasRemaining` check, and then to update the tests accordingly so we explicitly allow trailing bytes in a payload.

Let me know if this makes sense, we can talk about other approaches if you want. Sorry for the bad suggestion earlier -- I wasn't aware that the shim code was so permissive.

Comment (djb15) [reply node/pkg/watchers/solana/shim.go]: Appreciate you reviewing this change carefully, good catch! Claude seems to agree here, I've made the requested change. Decided to use a new wrapper function rather than adding an argument to the existing function to make it extra clear what's happening at the call site.

Also have some mainly LLM generated comments on the test cases, but I thought it would be helpful for future us to understand why we have the differing decoding cases.

Comment (djb15) [reply node/pkg/watchers/solana/shim.go]: Will deal with the rebase in a bit

Comment (johnsaigle) [reply node/pkg/watchers/solana/shim.go]: Awesome, yes the changes look good and I think the comments are super helpful. 👍🏻 I can approve after the rebase

**Review by johnsaigle — APPROVED** [review]

**Review by bemic — APPROVED** [review]
I found no code-level blocker.
